File: scripts/lib/constantsExtractor.py
from pathlib import Path
import json


def getVersionNumber():
    constantDict = __getSharedConstantsData()
    return constantDict["version"]


def getExtensionId():
    constantDict = __getSharedConstantsData()
    return constantDict["ids"]["extension"]


def getNativeAppId():
    constantDict = __getSharedConstantsData()
    return constantDict["ids"]["nativeApp"]


def __getSharedConstantsData():
    sharedConstPath = __getSharedConstantsPath()
    with open(sharedConstPath, "r") as f:
        constantDict = json.load(f)
        return constantDict


def __getSharedConstantsPath():
    rootPkgDir = __getRootPackagesDir()
    sharedConstantsPath = __findSharedConstantsJson(rootPkgDir)

    return sharedConstantsPath


def __findSharedConstantsJson(rootPkgDir):
    matchesGenerator = rootPkgDir.glob("**/sharedConstants.json")
    matchesList = list(matchesGenerator)

    if len(matchesList) == 0:
        raise FileNotFoundError("No shared constants file could be found.")
    elif len(matchesList) > 1:
        raise FileNotFoundError(
            (
                "Two or more copies of the share constants file were found."
                " There should only be 1"
            )
        )
    else:
        return matchesList[0]


def __getRootPackagesDir():
    pkgDirParts = __getPackageDir().parts
    index = pkgDirParts.index("packages")
    rootPkgDir = Path(*pkgDirParts[: index + 1])

    return rootPkgDir


def __getPackageDir():
    taskDir = Path(__file__).parent
    pkgDir = Path(
        *["packages" if part == "scripts" else part for part in taskDir.parts]
    )
    return pkgDir

File: scripts/extension/constantsExtractor.js
const path = require("path");
const glob = require("glob");
const jsonfile = require("jsonfile");

async function getVersionNumber() {
  const constantsFilePath = await __findPathToConstantsFile();

  return __getVersionFromFile(constantsFilePath);
}

async function getExtensionId() {
  const constantsFilePath = await __findPathToConstantsFile();

  return __getExtensionIdFromFile(constantsFilePath);
}

async function getNativeAppId() {
  const constantsFilePath = await __findPathToConstantsFile();

  return __getNativeAppIdFromFile(constantsFilePath);
}

async function __findPathToConstantsFile() {
  const sourceDir = __findRootPackageDirectory().abs;

  return __findSharedConstantsJson(sourceDir);
}

function __findRootPackageDirectory() {
  const scriptsDirParts = __dirname.split(path.sep);
  const index = scriptsDirParts.indexOf("scripts");
  if (index < 0) {
    throw new Error("Unable to find source directory from scripts directory");
  }

  const sourceDirParts = scriptsDirParts.slice(0, index + 1);
  sourceDirParts[index] = "packages";

  const sourceDirAbsPath = path.join(...sourceDirParts);
  const sourceDirRelPath = path.relative(__dirname, sourceDirAbsPath);

  return {
    abs: sourceDirAbsPath,
    rel: sourceDirRelPath,
  };
}

async function __findSharedConstantsJson(sourceDir) {
  return new Promise(function (resolve, reject) {
    const opts = {
      cwd: sourceDir,
      nodir: true,
      realpath: true,
    };

    glob("**/sharedConstants.json", opts, function (err, files) {
      if (err) {
        reject(err);
      } else {
        if (!files || files.length <= 0) {
          reject("No shared constants file could be found.");
        } else if (files.length > 1) {
          reject(
            "Two or more copies of the share constants file were found. There should only be 1"
          );
        } else {
          resolve(files[0]);
        }
      }
    });
  });
}

async function __getVersionFromFile(filepath) {
  const constantsDict = await __getConstantsDict(filepath);

  return constantsDict["version"];
}

async function __getExtensionIdFromFile(filepath) {
  const constantsDict = await __getConstantsDict(filepath);

  return constantsDict["ids"]["extension"];
}

async function __getNativeAppIdFromFile(filepath) {
  const constantsDict = await __getConstantsDict(filepath);

  return constantsDict["ids"]["nativeApp"];
}

async function __getConstantsDict(filepath) {
  return jsonfile.readFile(filepath);
}

module.exports.getVersionNumber = getVersionNumber;
module.exports.getExtensionId = getExtensionId;
module.exports.getNativeAppId = getNativeAppId;
